Introduce `RUSTC_STATIC_CLANG_RT_PATH` and `RUSTC_STATIC_UNWIND_PATH` envs

catap opened the pull request:
The goal of this envirements variable is easy statically link `libclang_rt` or `libunwind` into rust.

It also introduces a way to easy extend rust build in a way which allows to inject more static builds for some rare system if it required.

It haven't been documented because it seems as very deep hack and the user who needs it will discover this hack by reading sources ;)

Closes: https://github.com/rust-lang/rust/issues/59164

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @davidtwco (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4505034936?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c5398301bb7520d65ab0e3696593d7d800791f47/checks/4505034936/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_llvm/build.rs at line 2:
 use std::path::{Path, PathBuf};
 
 
-use build_helper::{output, tracked_env_var_os, maybe_static_clang_rt};
+use build_helper::{maybe_static_clang_rt, output, tracked_env_var_os};
 
 fn detect_llvm_link() -> (&'static str, &'static str) {
     // Force the link mode we want, preferring static by default, but
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_const_eval/src/transform/check_consts/post_drop_elaboration.rs" "/checkout/compiler/rustc_parse_format/src/tests.rs" "/checkout/compiler/rustc_const_eval/src/transform/check_consts/mod.rs" "/checkout/compiler/rustc_llvm/build.rs" "/checkout/compiler/rustc_const_eval/src/transform/check_consts/ops.rs" "/checkout/compiler/rustc_const_eval/src/transform/check_consts/qualifs.rs" "/checkout/compiler/rustc_const_eval/src/transform/check_consts/resolver.rs" "/checkout/compiler/rustc_parse_format/src/lib.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4505939586?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7df5d0204f2c71c5222db29f2e7da73a232d26aa/checks/4505939586/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
   Compiling libc v0.2.108
   Compiling memchr v2.4.1
   Compiling std v0.0.0 (/checkout/library/std)
error[E0432]: unresolved import `build_helper::maybe_static_clang_rt`
  |
  |
3 | use build_helper::maybe_static_clang_rt;
  |     |             |
  |     |             |
  |     |             help: a similar name exists in the module: `maybe_static_library`
  |     no `maybe_static_clang_rt` in the root
For more information about this error, try `rustc --explain E0432`.
error: could not compile `std` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4506061988?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/0952de6eef017d2a3c045445a398d1f38f584ab8/checks/4506061988/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling memchr v2.4.1
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.55
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0432]: unresolved import `build_helper::maybe_static_clang_rt`
  |
  |
3 | use build_helper::maybe_static_clang_rt;
  |     |             |
  |     |             |
  |     |             help: a similar name exists in the module: `maybe_static_library`
  |     no `maybe_static_clang_rt` in the root
For more information about this error, try `rustc --explain E0432`.
error: could not compile `unwind` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

catap committed:
Introduce `RUSTC_STATIC_CLANG_RT_PATH` and `RUSTC_STATIC_UNWIND_PATH` envs

The goal of this envirements variable is easy statically link
`libclang_rt` or `libunwind` into rust.

It also introduces a way to easy extend rust build in a way which allows
to inject more static builds for some rare system if it required.

It haven't been documented because it seems as very deep hack and the
user who needs it will discover this hack by reading sources ;)

davidtwco commented:
This probably warrants a [major change proposal](https://forge.rust-lang.org/compiler/mcp.html) as it adds a public-facing change (we'd have to support this environment variable going forwards).

catap commented:
@davidtwco not sure to be honest. This is very low-level hack that allows to inject some dependency when it needed and can't be discovered by cross building.

Another approach is adding (a lot) `if`'s inside `build.rs` or start to use such hack.

From my point of view this hack is the best possible way because if I rebuild it to target platform where `libunwind` isn't discovered automatically the first place to start digging how to hack it (for me) is `build.rs` for `library/unwind` where I've discovered it and enjoyed it :)

But, still, I may not understand and estimate right the long term cost of supporting it but from my point of view it nothing. Am I wrong?

davidtwco commented:
I guess its more rustc developer/distribution-facing than user-facing, @Mark-Simulacrum do you know how we decide to add environment variables to `build.rs` for std/rustc?

Mark-Simulacrum commented:
I agree this is distro/developer-facing. The requirement for adding this is mainly convincing someone, typically me or someone else familiar with our linking story (petrochenkov, often).

In this case, I'm not sure this is the right next step here -- can you elaborate on the motivation here? #59164 seems like a case where the library/crate being compiled should be including the library, rather than pushing for std to link it in.

libunwind already has llvm-libunwind option which, on some platforms, leads to a static link. clang_rt I don't think currently has such an option, but it seems plausible we could add it (rather than having an env variable ad-hoc providing that).

JohnCSimon commented:
Ping from triage: 
@catap can you please address the comment 
> In this case, I'm not sure this is the right next step here -- can you elaborate on the motivation here?

catap commented:
@JohnCSimon, @Mark-Simulacrum the root cause of motivation for this changes is attempt to build rust which can be run on macOS before 10.7.

Anyway, it still requires a few things, and right now I haven't got any time to finish this build.

If it fine, keep this PR open for a while and as soon as I achived that, I'll update/rebase this one and open a few more which is required.

bors commented:
:umbrella: The latest upstream changes (presumably #94480) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

JohnCSimon commented:
@Mark-Simulacrum - @catap has replied to your question

Mark-Simulacrum commented:
I think if we're going to add configuration like this, it should go through config.toml and into the build scripts via rustbuild-set environment variables (or other means of passing configuration, perhaps existing ones).

For libunwind, it seems like we already try to support static linkage via crt-static and/or llvm-libunwind=in-tree. Maybe those options suffice here? If not, maybe this configuration should be gated on those rather than *new* configuration options?

Dylan-DPC commented:
@catap any update on this pr? thanks

JohnCSimon commented:
@catap 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

barracuda156 commented:
@catap Kirill, could you re-visit this?
